library/core/src/iter/traits/iterator.rs

```diff
@@ -3613,29 +3595,26 @@ pub trait Iterator {
     /// assert!(xs.iter().eq_by(&ys, |&x, &y| x * x == y));
     /// ```
     #[unstable(feature = "iter_order_by", issue = "64295")]
-    fn eq_by<I, F>(mut self, other: I, mut eq: F) -> bool
+    fn eq_by<I, F>(self, other: I, eq: F) -> bool
     where
         Self: Sized,
         I: IntoIterator,
         F: FnMut(Self::Item, I::Item) -> bool,
     {
-        let mut other = other.into_iter();
-
-        loop {
-            let x = match self.next() {
-                None => return other.next().is_none(),
-                Some(val) => val,
-            };
-
-            let y = match other.next() {
-                None => return false,
-                Some(val) => val,
-            };
-
-            if !eq(x, y) {
-                return false;
+        #[inline]
+        fn compare<X, Y, F>(mut eq: F) -> impl FnMut(X, Y) -> ControlFlow<()>
+        where
+            F: FnMut(X, Y) -> bool,
+        {
+            move |x, y| {
+                if eq(x, y) { ControlFlow::CONTINUE } else { ControlFlow::BREAK }
             }
         }
+
+        match iter_compare(self, other.into_iter(), compare(eq)) {
```

Comment:
For example, instead of needing `iter_compare` here, I'm thinking that this could perhaps be
```rust
self.zip(other).try_for_each(compare) == ControlFlow::CONTINUE
```
using almost exactly the same `compare` you already have, just with a couple more `()`s,
maybe
```rust
        fn compare<X, Y, F>(mut eq: F) -> impl FnMut(X, Y) -> ControlFlow<()>
        where
            F: FnMut((X, Y)) -> bool,
        {
            move |(x, y)| {
                if eq(x, y) { ControlFlow::CONTINUE } else { ControlFlow::BREAK }
            }
        }
```

(Or whatever other phrasing of this you'd prefer -- `find_map` might be able to do it elegantly too, for example.)

@rustbot author

Reply:
I discarded the idea of using `zip` because it throws away the crucial information of which iterator is longer — even when zipping `&mut I`, `Zip` won't tell you which iterator ran out of elements. I believe that `zip_longest` is what we'd need here to implement this correctly and without overload.

You _could_ do something like `.map(Some).chain(once(None))` to both iterators and then zip those, but it feels like that would just add extra branches.

Reply:
Oh, you're right 🤦  Yeah, lexicographical can't just use `zip`.  Let me take another look at this.

@rustbot ready